cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,266 @@
+* [R2] Keep the specified amount of extra capacity at minimum
+  * If we allowed users to annotate their k8s pods so that they are mirrored to equivalent placeholder pods like https://github.com/kubernetes/autoscaler/pull/77#discussion_r117084488,
+    CA may actually produce much more extra capacity than specified by the user.
+
+    Let's say an user specified CA to reserve 10% extra capacity, but also marked a pod which requests cpu and memory more than 10% of current capacity. CA will result in creating a placeholder pod which consumes resource more than 10%.
+    So, the "10%" in this case is just a minimum requirement. CA may reserve more if the user instructed CA to do so by annotations. That's why I wrote `at minimum`. It is impossible to specify CA an exact rate of extra capacity when we want to suppoprt [R6] and [R7].
+* [R3] Ensure full compatibility with scale down
+* [R4] Only expose necessary and sufficient knobs to users
+
+Version 2
+
+contains [R1]-[R4] and also what is below:
+
+* [R5] Don't add unnecessarily large node
+  * A node is unnecessarily large when:
+    * Any pod that is intended to be scheduled faster by resource slack fit within a smaller node
+* [R6] Possible to provide extra space even for the "biggest" pod in term of memory and cpu
+  * However, if the biggest pod in the cluster was a database, we won't want to keep a hot standby node for the pod to move to
+* [R7] Possible to provide extra space even for every pod which is not biggest but do request some cpu(< max(pod_request_cpu)) and memory(< max(pod_request_mmory))
+  * Say, the "biggest pods" request 1000m cpu/1G mem, 10m cpu/10G mem respectively
+  * Even if they could fit within existing free space, a 500m(<1000m cpu)/5G mem pod has chances not to fit any node 
+* [R8] Reserve space even for the pods having anti pod affinity
+* [R9] Reserve space even for the pods having node selector/node affinity
+
+Version 3
+
+contains [R1]-[R9] and also what is below:
+
+* [R10] Works on considerably large clusters
+  * To move CA from beta to GA we need to handle 5000 nodes and 150 000 pods.
+ 
+### Specifications
+
+Version 1
+
+* [S1] TBD [R1-1]
+* [S2] Retain placeholder pods among iterations of CA and remove/re-assign/add only necessary placeholder pods. [R1-2]
+  * More concretely, for each CA iteration:
+  * Step 1: Calculate a `list of pushed-out placeholder pods`
+    * Iterate over existing placeholder pods to determine which one is "pushed out of the node" by actual pod(s)
+       * Iterate over ready nodes and see if `sum(pod requested cpu/mem)` exceeds a node's capacity. If it exceeds, push out placeholders until the sum becomes less than or equal to the node capacity.
```

Comment:
I think it's more consistent to run scheduler predicates, same as everywhere else (they check if there is enough cpu/mem amongst other things).